include/seqan3/range/container/concatenated_sequences.hpp

```diff
@@ -182,33 +182,73 @@ class concatenated_sequences
      * \brief Static constexpr variables that emulate/encapsulate seqan3::compatible (which doesn't work for types during their definition).
      * \{
      */
+    //!\cond
+    // unfortunately we cannot specialise the variable template so we have to add an auxiliary here
+    template <std::ranges::range t>
+        requires std::convertible_to<std::ranges::range_reference_t<t>, std::ranges::range_value_t<value_type>>
+    static constexpr bool is_compatible_with_value_type_aux = dimension_v<t> == dimension_v<value_type>;
+    //!\endcond
+
     //!\brief Whether a type satisfies seqan3::compatible with this class's `value_type` or `reference` type.
     //!\hideinitializer
     // we explicitly check same-ness, because these types may not be fully resolved, yet
-    template <typename t>
-    static constexpr bool is_compatible_value = std::is_same_v<remove_cvref_t<t>, value_type>       ||
-                                                std::is_same_v<remove_cvref_t<t>, reference>        ||
-                                                std::is_same_v<remove_cvref_t<t>, const_reference>  ||
-                                                (dimension_v<t> == dimension_v<value_type> &&
-                                                std::convertible_to<reference_t<t>, value_type_t<value_type>>);
-    //!\}
+    template <std::ranges::range t>
+    static constexpr bool is_compatible_with_value_type =
+        std::is_same_v<remove_cvref_t<t>, value_type> ||
+        std::is_same_v<remove_cvref_t<t>, reference> ||
+        std::is_same_v<remove_cvref_t<t>, const_reference> ||
+        (
+            !std::same_as<remove_cvref_t<t>, iterator> &&
```

Comment:
is this not contradicting the requirement of t in the top-level expression? In other words, can it ever be a iterator here?

Reply:
that's the problem here, if you are within an evaluation during the definition of this class.

That means the C++ parser started 

```c++
template <typename inner_type,
          typename data_delimiters_type = std::vector<typename inner_type::size_type>>
//!\cond
    requires reservible_container<std::remove_reference_t<inner_type>> &&
             reservible_container<std::remove_reference_t<data_delimiters_type>> &&
             std::is_same_v<size_type_t<inner_type>, value_type_t<data_delimiters_type>>
//!\endcond
class concatenated_sequences
{
```
and it did not reach

```c++
} // closing brace of concatenated_sequences
```

concatenated_sequences and all it's member are in a weird limbo state there not everything might be completely defined yet, e.g. iterator might be an incomplete type. The standard behaviour of gcc < 10 is to just assume that the concept check on such a type is true and tries to instantiate all the expressions in this static bool (you could comment that `std::ranges::range t` thing out and it would have the same behaviour). The static bool forces the expression to be able to short-circut and `std::same_as` is able to compare with incomplete types.

Like I wrote, I'm not completely sure why this works, but I have seen that the requires has no effect.

Reply:
That sounds definitely like a to complicated design, where the benefits seem to be mariginal.

Reply:
Yeah :(